### ISSUE TITLE
[Example Script] Is it possible to finetune VLM on the assistant only text instead of the entire sequence?

### ISSUE THREAD

**Issue by xhluca** [huggingface/trl#3751]
In the example script below, we learn how to SFT a VLM with trl. It computes the loss for the entire sequence, including generating the prompt and user content:

https://github.com/huggingface/trl/blob/5787f3bf6365beb0054f984055e4461394933935/examples/scripts/sft_vlm.py#L86-L103

Is there a reason why we don't mask the system prompt and user text with `-100` tokens, similar to how we mask images? If we want to apply such mask, what would be the minimal change to bring to this script to achieve the desired outcome?

**Comment by pramodith** [comment]
@xhluca doesn't setting the `assistant_only_loss` flag in the SFTConfig's initialization args work for you? https://huggingface.co/docs/trl/main/en/sft_trainer#train-on-assistant-messages-only

**Comment by xhluca** [comment]
It only applies to a format that's not compatible with the processor output in the example. So it should not work (I did try but I don't think the flag makes a difference).

**Comment by pramodith** [comment]
Thanks for the info! I dove a bit deeper into this and unfortunately there's no easy way to get the `assistant_tokens_mask` for multimodal models.

Currently the `AutoProcessor`'s `apply_chat_template` function does accept an argument called `return_assistant_tokens_mask`, but this only works when the chat template of the model has `{ %generation%}` and `{%endgeneration%}` block in the chat template.

I tried providing a custom template that includes those tags to SmolVLM but `apply_chat_template` still won't work because we need to tokenize the input messages for the `assistant_tokens_mask `to be returned but with a multimodal dataset we can't tokenize it before applying the chat template, because we need to identify where to insert the images 🙃.

The easiest route would be to call a custom function that identifies all the token spans between <|im_start|>assistant and <|im_end|>inside the collator and then do something like ` labels[assistant_mask.bool()]=-100`.

@qgallouedec I think it'd be great for trl to have some sort of utility function to help with this. Essentially a function that returns the assistant_mask in cases where the default chat template doesn't have the {%generation%} block. I can work on this if you think this is a good idea.

**Comment by xhluca** [comment]
seems like a catch 22! would be great to have first-party support for that. i think a lot of chat lm scenarios have become vlm chat scenarios (since most chat models can receive images as input now), so it would probably benefit a lot of users!

**Comment by weathon** [comment]
Same, assistant_only_loss throw an error

**Comment by zwverse** [comment]
plus 1.  having assistant_only_loss support on VLLM would be really helpful.   We are running into a use case which needs the assistant_only_loss

**Comment by qgallouedec** [comment]
If your data is single-turn (one user message, one assistant message), one alternative is just to split into to columns: `"prompt"` and `"completion"`. And the `SFTTrainer` will compute the loss on the completion part only, which will be equivalent to computing the loss on assistant only.

**Comment by xhluca** [comment]
@qgallouedec could you share an example please? i'm not sure i'm following

**Comment by qgallouedec** [comment]
You can use a dataset formatted like this:

https://huggingface.co/datasets/trl-internal-testing/zen-image/viewer/conversational_prompt_completion

And when you run

```python
from trl import SFTTrainer
from datasets import load_dataset

dataset = load_dataset("trl-internal-testing/zen-image", "conversational_prompt_completion", split="train")

trainer = SFTTrainer(
    model="Qwen/Qwen3-VL-2B-Instruct",
    train_dataset=dataset,
)
trainer.train()
```


the loss will be computed on the completion column only, which in this case is the assistant part.

**Comment by albertvillanova** [comment]
Hey @xhluca, sorry for the long silence here — we're catching up on the backlog. 

Good news: this is now actively tracked in #5471, which adds the `{% generation %}` training chat templates that VLM `assistant_only_loss` needs — Gemma3, Qwen3-VL and Idefics3 have already landed, with Qwen2-VL/2.5-VL, SmolVLM and LLaVA still open. VLM `assistant_only_loss` is intentionally guarded until those templates are ready (that's the error @weathon and @zwverse hit: https://github.com/huggingface/trl/blob/fceadc75c81bd103a6cf7f870eafddf1106b6065/trl/trainer/sft_trainer.py#L1020-L1024). In the meantime, for single-turn data you can get assistant-only loss today by formatting your dataset with separate `prompt`/`completion` columns — `SFTTrainer` then computes the loss on the completion only (there's no VLM guard on `completion_only_loss`, and no config change needed). 

I'll close this in favor of #5471 to keep everything in one place — please follow/+1 there. Thanks for the detailed thread!

**Comment by albertvillanova** [comment]
Assistant-only loss for vision datasets is now tracked in #7376, with the implementation in #7377 (requires `transformers>=5.18.0`).
